### ISSUE TITLE
[BR]: How to Configure Fail2ban Rules for Multiple Ports

### ISSUE THREAD

**Issue by onlineyh** [fail2ban/fail2ban#3572]
### Environment:
ubuntu 20.04
apt install fail2ban -y

- Fail2Ban version  : 
- OS, including release name/version : 
- [ ] Fail2Ban installed via OS/distribution mechanisms
- [ ] You have not applied any additional foreign patches to the codebase
- [ ] Some customizations were done to the configuration (provide details below is so)

#### Steps to reproduce
[mstsc-nginx]

enabled = true
port = 6666,7777
filter = mstsc
logpath = /var/log/nginx/tcp-access.log
action = iptables[name=MSTSC_NGINX, port=6666,7777, protocol=tcp]
maxretry = 3
findtime = 360
bantime = 86400
#innoreip = 192.168.16.248

#### Observed behavior
I found that the blacklist IP cannot access port 6666 and can access port 7777. I don't know where the mistake is
#### Any customizations done to /etc/fail2ban/ configuration

```
```

#### Relevant parts of /var/log/fail2ban.log file:

```
```

#### Relevant lines from monitored log files:

```
Chain f2b-MSTSC_NGINX (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 REJECT     all  --  *      *       83.97.73.119         0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 REJECT     all  --  *      *       60.16.5.243          0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 REJECT     all  --  *      *       36.139.81.127        0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 REJECT     all  --  *      *       223.104.68.108       0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 REJECT     all  --  *      *       192.168.16.2         0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 REJECT     all  --  *      *       141.98.11.58         0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 REJECT     all  --  *      *       139.9.179.213        0.0.0.0/0            reject-with icmp-port-unreachable
    0     0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0           

```

**Comment by sebres** [comment]
You have to use multiport action, and `iptables` uses single port by default:
https://github.com/fail2ban/fail2ban/blob/f8f8c046a277fbf17fcc0a3075234adb07274dcf/config/action.d/iptables.conf#L13

Normally, fail2ban has a definition `banaction`, which is set in `jail.conf`:
https://github.com/fail2ban/fail2ban/blob/f8f8c046a277fbf17fcc0a3075234adb07274dcf/config/jail.conf#L208-L212
If it is also valid for your package, simply don't overwrite parameter `action`, it'd use multiport `banaction` by default.
Moreover related to your iptables excerpt it looks like some override made allports action by default.

Anyway, your config is not correct for multiple ports, so either do this:
```ini
port = 6666,7777
banaction = iptables[type=multiport]
```
or that:
```ini
action = iptables[type=multiport, name=MSTSC_NGINX, port=6666,7777, protocol=tcp]
```

And restart (don't reload) the jail or fail2ban hereafter.

By the way, better would be to switch to `iptables-ipset` action, if ipset is available and supported by your kernel (it'd create only one rule in INPUT chain and IPs going to ipset hash), or probably to `nftables` action, also if it is supported.

**Comment by onlineyh** [comment]
Thank you! Configuration is valid!

**Comment by onlineyh** [comment]
May I ask how to use Fail2ban to call shell scripts or Python scripts to send blacklisted IPs?
 I want to use a Python script to send the blacklisted IP addresses to my phone

**Comment by sebres** [comment]
> May I ask how to use Fail2ban to call shell scripts or Python scripts to send blacklisted IPs?

Simply create an action doing that "send to phone", see for example any mail action (or [dummy](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/dummy.conf) action), here the simplest one, let say `action.d/my-action.conf`:
```ini
[Definition]
# bypass ban for restored tickets
norestored = 1
actionban = send-to-phone.sh "<token>" "<ip>"
actionunban =
actionflush = true
[Init]
token = 
```

And then add it to the `action` parameter to jail or default section:
```ini
[DEFAULT]
action = %(action_)s
         my-action[token="012345-6789-abcdef"]
```
(Note that all actions invoked one after the other per each single ban, so the next bans may be retarded due to execution time of `my-action`, so better if you'd implement it as fast as possible, e. g. asynchronously).

As for python - it is possible to create pythonic actions - [here](https://github.com/fail2ban/fail2ban/blob/master/config/action.d/smtp.py) is an example for smtp action completely written in python.
Basically something like that could be enough:
```python
from fail2ban.server.actions import ActionBase
class MyAction(ActionBase):
  def __init__(
    self, jail, name, token
  ):
    self._token = token
    super(MyAction, self).__init__(jail, name)
    # bypass ban for restored tickets
    self.norestored = 1

  def ban(self, aInfo):
    # if aInfo.get('restored'): return
    do_send_to_phone(self._token, aInfo["ip"])

**Comment by onlineyh** [comment]
Thank you! I have successfully configured it!

**Comment by onlineyh** [comment]
How can I increase the number of ports,
[mstsc-nginx]
enabled = true
port = 7777,8888,16360,15207,15690,15590,15770,15410,15940,15880,16116,15700,15200,15133,15134
filter = mstsc
logpath = /var/log/nginx/tcp-access.log
banaction = iptables-multiport
banaction_allports = iptables-allports
action = %(banaction)s[port="%(port)s", protocol=tcp, chain="%(chain)s"]
         send-mstsc[token="9025"]
#action = iptables[type=multiport,name=MSTSC_NGINX, port=6666,7777,8888, protocol=tcp]
maxretry = 3
findtime = 360
bantime = 2592000
After adding an additional port, fail2ban cannot be started

**Comment by onlineyh** [comment]
Is it possible that I have added too many ports, causing fail2ban to fail to start
How many ports can be added by default

**Comment by onlineyh** [comment]
Thank you! It has been resolved according to your suggestion!

**Comment by onlineyh** [comment]
I still want to inquire about how to configure ipset
ipset and iptable
I am more familiar with iptable
